bootstrap/libtest: print test name eagerly on failure even with `verbose-tests=false` / `--quiet`

jyn514 opened the pull request:
Previously, libtest would wait until all tests finished running to print the progress, which made it
annoying to run many tests at once (since you don't know which have failed). Change it to print the
names as soon as they fail.

This makes it much easier to know which test failed without having to wait for compiletest to completely finish running. Before:
```
Testing stage0 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)

running 15274 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii    88/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   176/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   264/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   352/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   440/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   528/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiFFiiiiiii
...
```

After:
```
Testing stage0 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)

running 15274 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii    88/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   176/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   264/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   352/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   440/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   528/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
[ui] tests/ui/associated-type-bounds/implied-in-supertrait.rs ... F

[ui] tests/ui/associated-type-bounds/return-type-notation/basic.rs#next_with ... F
iiiiiiiiiiiii
...
```

This serves a similar use case to the existing RUSTC_TEST_FAIL_FAST, but is on by default and as a result much more discoverable. We should consider unifying RUSTC_TEST_FAIL_FAST with the `--no-fail-fast` flag in the future for consistency and discoverability.

jyn514 committed:
bootstrap: print test name on failure even with `verbose-tests=false`

This makes it much easier to know which test failed without having to wait for compiletest to completely finish running. Before:
```
Testing stage0 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)

running 15274 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii    88/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   176/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   264/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   352/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   440/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   528/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiFFiiiiiii
...
```

After:
```
Testing stage0 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)

running 15274 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii    88/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   176/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   264/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   352/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   440/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii   528/15274
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
[ui] tests/ui/associated-type-bounds/implied-in-supertrait.rs ... F

[ui] tests/ui/associated-type-bounds/return-type-notation/basic.rs#next_with ... F
iiiiiiiiiiiii
...
```

This serves a similar use case to the existing RUSTC_TEST_FAIL_FAST, but is on by default and as a result much more discoverable.
We should consider unifying RUSTC_TEST_FAIL_FAST with the `--no-fail-fast` flag in the future for consistency and discoverability.

Kobzol commented:
This is great! It has been bugging me that the failures are not shown until the end. I had no idea about `RUSTC_TEST_FAIL_FAST`, of course.

Kobzol commented:
I tried it locally and the immediate feedback is great, but it also really borks the nice table output of compiletest :/ In pytest, there's either the succint `...FF....` output, or a more verbose output like this:
```
test 1..... [PASSED]
test 2..... [FAILED]
test 3..... [PASSED]
```
But probably since we have so many tests, it would print too many lines and the failed tests would very quickly "scroll up".

jyn514 commented:
you can try out the more verbose output locally if you like, `verbose-tests = true`

Kobzol commented:
With verbose tests, the failed tests indeed disappear too quickly I think. So this new output mode is definitely useful, to keep succint output, but also warn about test failures.

Mark-Simulacrum commented:
Does this mean we're diverging from the libtest formatting? Or does libtest not have a "just dots" output?

I would prefer to avoid diverging if libtest *does* have such an output (i.e., maybe we can update that, since this seems like a reasonable change), but if this is already custom bootstrap code then r=me.

@rustbot author

jyn514 commented:
https://github.com/rust-lang/team/pull/1051

jyn514 committed:
libtest: Print the names of failed tests eagerly

Previously, libtest would wait until all tests finished running to print the progress, which made it
annoying to run many tests at once (since you don't know which have failed). Change it to print the
names as soon as they fail.

This also adds a test for the terse output; previously it was untested.

jyn514 commented:
> Does this mean we're diverging from the libtest formatting? Or does libtest not have a "just dots" output?

yes. i updated this to update libtest too.

jyn514 commented:
@rustbot ready

jyn514 commented:
r? libs

cuviper commented on library/test/src/formatters/terse.rs:
```diff
@@ -47,8 +47,12 @@ impl<T: Write> TerseFormatter<T> {
+        // to see which test failed without having to wait for them all to run.
+        self.write_plain(format!("\n{name} --- "))?;
+        self.write_pretty("F", term::color::RED)?;
+        self.write_progress()
```
I think this might want to print progress *first*, padded to the right alignment, because it looks awkward to me to have "count/total" on the same line as a single failure. But this could be debated.

Maybe we should also spell out `FAILED` here like the non-terse case?

jyn514 replied:
> I think this might want to print progress _first_, padded to the right alignment, because it looks awkward to me to have "count/total" on the same line as a single failure. But this could be debated.

:+1: i agree this looks better.

i don't feel strongly about F vs FAILED; i kind of like that F is shorter, since this is the terse format, but also it's already getting its own line so it's not that much shorter relatively. i've changed it to FAILED.

cuviper commented on library/test/src/formatters/terse.rs:
```diff
@@ -65,15 +69,21 @@ impl<T: Write> TerseFormatter<T> {
+    }
+
+    fn write_progress(&mut self) -> io::Result<()> {
+        let out = format!(" {}/{}\n", self.test_column, self.total_test_count);
```
We still need a `test_count` instead of using the column here, or else we're missing overall progress:

```
........................................................................................ 88/1000
........................................................................................ 88/1000
........................................................................................ 88/1000
...
```

jyn514 replied:
good catch, thank you! i added a test which catches this bug and also a couple others i ran into while fixing it.

cuviper reviewed (commented):
cc @rust-lang/testing-devex for `test` crate changes

epage commented:
No strong preference for myself.

Personally, I'm mixed on this but I see this as a short-lived problem.  The testing team's first focus is on json.  With that, I hope to move rendering to `cargo test` and de-emphasize the format modes within libtest (to simplify expectations on custom test harnesses).  We can then play with what works well for a rich, easy to read experience for users with `cargo test` and use fancier terminal features.

jyn514 commented:
bump @cuviper :)

bors commented:
:umbrella: The latest upstream changes (presumably #121400) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cuviper commented:
I think it's good, but needs a rebase.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8150060307/job/22275698792) [(plain)](https://github.com/rust-lang/rust/commit/9635b0a625dcefd75aa658df236cea2aaf07c201/checks/22275698792/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 writing image sha256:d8e1d2d41b86a5ca5d86faab6d1a16d440264516cc6414c81c5c93086f3813ed done
#12 naming to docker.io/library/rust-ci done
#12 DONE 9.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Tue Mar  5 02:18:30 UTC 2024
  network time: Tue, 05 Mar 2024 02:18:30 GMT
  network time: Tue, 05 Mar 2024 02:18:30 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling compiletest v0.0.0 (/checkout/src/tools/compiletest)
    Finished release [optimized] target(s) in 13.84s
##[endgroup]
##[group]Testing stage2 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)
error: detected legacy-style directives in compiletest test: /checkout/tests/ui/test-attrs/terse.rs:1, please use `ui_test`-style directives `//@` instead:Some(
        None,
        "compile-flags: --test",
    ),
)
)
errors encountered during EarlyProps parsing: /checkout/tests/ui/test-attrs/terse.rs
errors encountered during EarlyProps parsing
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:10:28
  local time: Tue Mar  5 02:29:35 UTC 2024
```

</details>

oli-obk commented:
@bors r=cuviper

bors commented:
:pushpin: Commit 8bfe9dbae20628bfcea390556c40a991984692ec has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 8bfe9dbae20628bfcea390556c40a991984692ec -->
<!-- homu: {"type":"Approved","sha":"8bfe9dbae20628bfcea390556c40a991984692ec","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
